### PR TITLE
fix(autograd): sinc_backward NaN for second derivative at x=0

### LINKED ISSUE #89459: Higher order derivatives of sinc explode
### 🐛 Describe the bug

First order derivatives of `torch.sinc(x)` are fine at `x=0`: they were previously reported as exploding in https://github.com/pytorch/pytorch/issues/56760 and https://github.com/pytorch/pytorch/issues/53989 and were fixed in https://github.com/pytorch/pytorch/pull/56763.

However, higher order derivatives or `torch.sinc(x)` at `x=0` still explode:

```python
zero = torch.zeros(1, requires_grad = True)
dx = torch.autograd.grad(torch.sinc(zero), zero, create_graph=True)[0]
torch.autograd.grad(dx, zero)[0]
tensor([nan])
```

By reading off the Taylor series of `sinc(x) = sin(x)/x = 1- x^2/3! + x^4/5! ...`, we get that the second derivative should have of limit of `-1/3`.

### Versions

PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5.1 (x86_64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.11)
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:38:35) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] torch==1.12.1
[conda] Could not collect

### PR DESCRIPTION
**Pull request by Jayantparashar10**
`sinc_backward` divided directly by `self_squared_pi`, which is zero at x=0. Since `sinc'(x) = (pi*x*cos(pi*x) - sin(pi*x)) / (pi*x^2)` has a 0/0 singularity there, NaN entered the autograd graph and sinc''(0) came out NaN instead of -pi^2/3.

The fix introduces a safe denominator (`ones_like` where `self_squared_pi == 0`) to block NaN in the general branch, and replaces the `scalar_tensor(0.0)` true-branch of the outer `at::where` with `grad * (-self_pi * M_PI / 3.0)` — the first Taylor term of `sinc'(x) = -pi^2*x/3`. This evaluates to 0 at x=0 (correct: `sinc'(0) = 0`) and lets autograd trace through it to recover `sinc''(0) = -pi^2/3`. Using a constant scalar as the true-branch blocks higher-order gradients from flowing through it entirely.

Fixes: #89459 

Test Plan:
- `python test/test_unary_ufuncs.py -k sinc` — 554 passed, 18 skipped
- `lintrunner -a torch/csrc/autograd/FunctionsManual.cpp` — no issues in changed lines

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -5367,10 +5367,16 @@ Tensor log1p_backward(const Tensor& grad, const Tensor& self) {
 Tensor sinc_backward(const Tensor& grad, const Tensor& self) {
   auto self_pi = self * M_PI;
   auto self_squared_pi = self * self * M_PI;
-  auto out = grad *
-      ((self_pi * self_pi.cos() - self_pi.sin()) / self_squared_pi).conj();
+  // Safe denom prevents NaN from entering the autograd graph at x=0;
+  // at::where alone does not block it from poisoning higher-order grads.
+  auto safe_denom = at::where(
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: We should really just simplify this with masked_fill

Comment (Jayantparashar10) [reply torch/csrc/autograd/FunctionsManual.cpp]: @Skylion007 Thanks for pointing this out! Simplified using `masked_fill` and also computed the mask once, reusing it for both the denominator fill and the final `at::where` to avoid the duplicate `self_squared_pi == 0.0` expression.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -5367,10 +5367,15 @@ Tensor log1p_backward(const Tensor& grad, const Tensor& self) {
...
-      self_squared_pi == 0.0, at::scalar_tensor(0.0, grad.options()), out);
+  auto mask = (self_squared_pi == 0.0);
+  // Prevent NaN from entering the graph at x=0; even though those positions
+  // are overridden below, 0*NaN = NaN during higher-order grad computation.
+  auto safe_denom = self_squared_pi.masked_fill(mask, 1.0);
+  auto out =
+      grad * ((self_pi * self_pi.cos() - self_pi.sin()) / safe_denom).conj();
+  // Use -pi^2*x/3 (first Taylor term of sinc') as the x=0 branch so that
+  // autograd can trace through it and recover sinc''(0) = -pi^2/3 correctly.
+  return at::where(mask, grad * (-self_pi * M_PI / 3.0).conj(), out);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Eagerly evaluate M_PI/3.0 in parens so it can be const folded before being broadcast to each torch tensor element.

Comment (Jayantparashar10) [reply torch/csrc/autograd/FunctionsManual.cpp]: Thanks for pointing this out! Resolved — also replaced `self * self * M_PI` with` self_pi * self` since `self_pi = self * M_PI` is already computed above, avoiding one redundant scalar broadcast.

**Comment by Jayantparashar10** [comment]
@drisspg @Skylion007 @albanD Please review this and if there any thing to update please let me know

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
@claude review this PR

**Comment by Skylion007** [comment]
Test coverage is the main problem here

**Comment by Skylion007** [comment]
@Jayantparashar10 please add some test coverage at x==0 for the proper derivative
